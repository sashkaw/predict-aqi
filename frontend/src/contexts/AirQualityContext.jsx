import { createContext, useContext, useState, useEffect } from "react";
import axios from "axios";

export const AQIContext = createContext();
export const useAQIContext = () => useContext(AQIContext);

export const AQIProvider = (props) => {
    const [site, setSite] = useState({
        "id": 1,
        "name": "Amazon Park - Eugene, OR",
        "latitude": 44.026280,
        "longitude": -123.083715,
      });

    const [currentData, setCurrentData] = useState(null);
    const [forecastData, setForecastData] = useState([]);
    const [futureTimeSteps, setFutureTimeSteps] = useState([]);

    // Fetch AQI data from backend after DOM has updated
    useEffect(() => {
        axios('/aqi/')
            .then((response) => {
                //console.log(response.data.aqi_current.at(-1))
                // Get most recent AQI observation
                setCurrentData(response.data.aqi_current.at(-1));
                // Get forecasted AQI for future time steps
                setForecastData(response.data.aqi_forecast);
                // Get corresponding time steps for forecasted data
                setFutureTimeSteps(response.data.future_timesteps);
                
            });
    }, [site]); // Only run if value of site changes
    //console.log(futureTimeSteps, currentData)

    return (
        <AQIContext.Provider
            value={{
                site, 
                setSite,
                currentData,
                setCurrentData,
                forecastData,
                setForecastData,
                futureTimeSteps,
                setFutureTimeSteps,
            }}
            >
                {props.children}
            </AQIContext.Provider>
    );

};



