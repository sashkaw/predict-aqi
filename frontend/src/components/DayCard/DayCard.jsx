import dayjs from "dayjs";
import { useAQIContext } from "../../contexts/AirQualityContext";

export function DayCard({ item, index, className, onClick }) {
    // Get current time step based on forecast data index
    const { futureTimeSteps } = useAQIContext();
    const cardTimeStep = futureTimeSteps[index];

    return (
        <li className={className} onClick={onClick}>
            <span className="timestep-name">{cardTimeStep}</span>
            <span className="timestep-aqi">{item}</span>
        </li>
    );
}