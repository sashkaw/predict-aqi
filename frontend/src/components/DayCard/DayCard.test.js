import renderer, { act } from "react-test-renderer";
import { getByRole, render, screen } from "@testing-library/react";
import { useEffect, useState, useContext } from "react";
import { DayCard } from "./DayCard";
import { AQIContext, useAQIContext } from "../../contexts/AirQualityContext";
import TestAQIProvider from "../Fixtures.js";
import { unmountComponentAtNode } from "react-dom";
import { toBeInTheDocument } from "@testing-library/jest-dom";

// Set up DayCard with mock data
const Fixture = () => {
    const Child = () => {
        const { forecastData, futureTimeSteps, setCurrentData, setCurrentTime } = useAQIContext();
        const [selectedCard, setSelectedCard] = useState(0);
        const index = 3;
        const item = forecastData[index];
        return(
            <DayCard
                className={index === selectedCard ? "active" : ""}
                onClick={() => {
                    setSelectedCard(index);
                    setCurrentData(item);
                    setCurrentTime(futureTimeSteps[index]);
                }}
                key={index}
                index={index}
                item={item}
            />
        )
    };
    // Return DayCard wrapped within Provider
    return (
        <TestAQIProvider>
            <Child />
        </TestAQIProvider>
    );
}

// Helper function for snapshot testing
const matches = (children) => expect(
    renderer.create(children).toJSON()
).toMatchSnapshot();

// Tests
describe("<DayCard />", () => {
    let container = null;
    beforeEach(() => {
        // Setup a DOM element as a render target
        container = document.createElement("div");
        document.body.appendChild(container);
    });
    afterEach(() => {
        // Cleanup on exiting
        unmountComponentAtNode(container);
        container.remove();
        container = null;
    });

    test("Renders DayCard element and text content", () => {
        const { getByRole } = render(<Fixture />, container);
        // Renders list item
        expect(getByRole("listitem")).toBeInTheDocument();
        // Renders forecast timestep and data
        expect(document.body.querySelector(".timestep-name").textContent).toBe("2pm");
        expect(document.body.querySelector(".timestep-aqi").textContent).toBe("AQI: 5");
        // Render matches snapshot
        matches(<Fixture />);
    });
})

