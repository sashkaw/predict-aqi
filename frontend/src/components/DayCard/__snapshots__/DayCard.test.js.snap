// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DayCard /> Renders DayCard element and text content 1`] = `
<li
  className=""
  onClick={[Function]}
>
  <span
    className="timestep-name"
  >
    2pm
  </span>
  <span
    className="timestep-aqi"
  >
    AQI: 
    5
  </span>
</li>
`;
